appinventor/components/src/com/google/appinventor/components/runtime/ChatBot.java

```diff
@@ -0,0 +1,488 @@
+    "BgkqhkiG9w0BAQsFAAOCAQEAIkKr3eIvwZO6a1Jsh3qXwveVnrqwxYvLw2IhTwNT\n" +
+    "/P6C5jbRnzUuDuzg5sEIpbBo/Bp3qIp7G5cdVOkIrqO7uCp6Kyc7d9lPsEe/cbF4\n" +
+    "aNwNmdWroRN1y0tuMU6+z7frd5pOeAZP9E/DM/0Uaz4yVzwnlvZUttaLymyMhH54\n" +
+    "isGQKbAqHDFtKZvb6DxsHzrO2YgeaBAtjeVhPWiv8BhzbOo9+hhZvYHYtoM2W+Ze\n" +
+    "DHuvv0v+qouphftDKVBp16N8Pk5WgabTXzV6VcNee92iwbWYDEv06+S3AF/q2TBe\n" +
+    "xxXtAa5ywbp6IRF37QuQChcYnOx7zIylYI1PIENfQFC2BA==\n" +
+    "-----END CERTIFICATE-----\n";
+
+  public static final String CHATBOT_SERVICE_URL =
+    "https://chatbot.appinventor.mit.edu/chat/v1";
+
+  private final Activity activity;
+  private String apiKey;
+  private static final String LOG_TAG = "ChatBot";
+  private static final boolean DEBUG = false;
+
+  private SSLSocketFactory sslSockFactory = null; // Socket Factory for using
+                                                  // SSL
+
+  private String system = "";   // The "System" string for ChatGPT
+  private String uuid = "";     // The UUID for continuing conversations
+
+  /**
+   * Creates a new component.
+   *
+   * @param container  container, component will be placed in
+   */
+  public ChatBot(ComponentContainer container) {
+    super(container.$form());
+    activity = container.$context();
+  }
+
+  @SimpleFunction(description = "Reset the current conversation, Chat bot will forget " +
+    "any previous conversation when resonding in the future.")
+  public void ResetConversation() {
+    this.uuid = "";
+  }
+
+  @SimpleFunction(description = "Ask a question of the Chat Bot. Successive calls will " +
+    "remember information from earlier in the conversation. Use the \"ResetConversation\" " +
+    "function to reset for a new conversation.")
+  public void Converse(final String question) {
+
+    AsynchUtil.runAsynchronously(new Runnable() {
+      @Override
+      public void run() {
+        performRequest(uuid, question);
+      }
+      });
+  }
+
+  private void performRequest(String uuid, String question) {
+    // languageToTransateTo is provided either as a two letter code, or two
+    // two letter codes separated by a dash. If only one two letter code is
+    // provided, it is the target language and we set the source language to auto
+    // which tells the service to detect the language
+    HttpsURLConnection connection = null;
+    ensureSslSockFactory();
+    try {
+      byte [] decodedToken = Base58Util.decode(apiKey);
+      ChatBotToken.token token = ChatBotToken.token.parseFrom(decodedToken);
+      ChatBotToken.request request;
+      if (system.equals("") || !uuid.equals("")) {  // No system string
+        request = ChatBotToken.request.newBuilder()
+          .setToken(token)
+          .setUuid(uuid)
+          .setQuestion(question).build();
+      } else {
+        request = ChatBotToken.request.newBuilder()
+          .setToken(token)
+          .setUuid(uuid)
+          .setSystem(system)
+          .setQuestion(question).build();
+      }
```

Comment:
This can be refactored using the builder paradigm so that all of the common properties (token, uuid, question) are set first and then we only set the system property on `uuid.equals("") && !system.equals("")`, rather than duplicating the code to set the common properties across both branches of the `if`.

Reply:
fixed